google/cloud/bigtable/row_merger.py

```diff
@@ -0,0 +1,498 @@
+        self.state = _State.ROW_START
+        self.row = None
+
+    def process_chunks(self, response):
+        """
+        Process the chunks in the given response and yield logical rows.
+        The internal state of this class will maintain state across multiple
+        response protos.
+        """
+        if response.last_scanned_row_key:
+            if self.last_seen_row_key >= response.last_scanned_row_key:
+                raise InvalidChunk("Last scanned row key is out of order")
+            self.last_seen_row_key = response.last_scanned_row_key
+
+        for chunk in response.chunks:
+            if chunk.reset_row:
+                self._handle_reset(chunk)
+                continue
+
+            if self.state == _State.ROW_START:
+                self._handle_row_start(chunk)
+
+            if self.state == _State.CELL_START:
+                self._handle_cell_start(chunk)
+
+            if self.state == _State.CELL_IN_PROGRESS:
+                self._handle_cell_in_progress(chunk)
+
+            if self.state == _State.CELL_COMPLETE:
+                self._handle_cell_complete(chunk)
+
+            if self.state == _State.ROW_COMPLETE:
+                yield self._handle_row_complete(chunk)
+            elif chunk.commit_row:
+                raise InvalidChunk(
+                    f"Chunk tried to commit row in wrong state (${self.state})"
+                )
+
+    def _handle_reset(self, chunk):
+        if self.state == _State.ROW_START:
+            raise InvalidChunk("Bare reset")
+        if chunk.row_key:
+            raise InvalidChunk("Reset chunk has a row key")
+        if chunk.HasField("family_name"):
+            raise InvalidChunk("Reset chunk has family_name")
+        if chunk.HasField("qualifier"):
+            raise InvalidChunk("Reset chunk has qualifier")
+        if chunk.timestamp_micros:
+            raise InvalidChunk("Reset chunk has a timestamp")
+        if chunk.labels:
+            raise InvalidChunk("Reset chunk has labels")
+        if chunk.value:
+            raise InvalidChunk("Reset chunk has a value")
+
+        self.state = _State.ROW_START
+        self.row = None
+
+    def _handle_row_start(self, chunk):
+        if not chunk.row_key:
+            raise InvalidChunk("New row is missing a row key")
+        if self.last_seen_row_key and self.last_seen_row_key >= chunk.row_key:
+            raise InvalidChunk("Out of order row keys")
+
+        self.row = _PartialRow(chunk.row_key)
+        self.state = _State.CELL_START
+
+    def _handle_cell_start(self, chunk):
+        # Ensure that all chunks after the first one either are missing a row
+        # key or the row is the same
+        if self.row.cells and chunk.row_key and chunk.row_key != self.row.row_key:
+            raise InvalidChunk("row key changed mid row")
+
+        if not self.row.cell:
+            self.row.cell = _PartialCell()
+
+        if chunk.HasField("family_name"):
```

Comment:
Can we simplify this to:

if !chunk.HasField("family_name") or !chunk.HasField("qualifier"):
  raise InvalidChunk("missing family_name or qualifier for a new cell")

?

Reply:
And should we move this check before we create a new PartialCell?

Reply:
Add a comment to explain the family and qualifier buffer